### PR TITLE
Translate strings in page config.

### PR DESCRIPTION
**Pull request by keyurva**
* Previously we added support for non-English queries [(PR)](https://github.com/datacommonsorg/website/pull/3650).
* However, the response was still in English.
* This PR translates the response strings to the queried language.

[translate-page-config.webm](https://github.com/datacommonsorg/website/assets/1221814/72125a9d-aae6-4405-9b9b-686950759254)

### CODE REVIEW

File: server/lib/translator.py [hunk 1]
```diff
@@ -21,42 +22,168 @@
...
+
+
+class TranslationStringTokens:
+  """Class that replaces tokens (like "${date}") in strings 
+with replacements that won't be translated (like "___") to send for translation.
+
+It also supports a method to reinsert tokens back into the translated string.
+"""
+
+  _TOKEN_PATTERN = r"\${[^}]+}"
```
Comment (pradh) [inline server/lib/translator.py:65]: comment what the pattern is?

File: server/lib/translator.py [hunk 2]
```diff
@@ -21,42 +22,168 @@
...
+  def reinsert_tokens(self, translation):
+    for token in self.tokens:
+      translation = re.sub(TranslationStringTokens._TOKEN_REPLACEMENT,
+                           token,
+                           translation,
+                           count=1)
+    return translation
+
+
+# Translates display strings in the provided page config and
```
Comment (pradh) [inline server/lib/translator.py:87]: There's a lot of logic here that could use some unit tests

File: server/lib/translator.py [hunk 3]
```diff
@@ -21,42 +22,168 @@
...
+    for block in category.get("blocks", []):
+      _maybe_add(block.get("title"))
+      for column in block.get("columns", []):
+        for tile in column.get("tiles", []):
+          _maybe_add(tile.get("description"))
+          _maybe_add(tile.get("title"))
+    for sv in category.get("statVarSpec", {}).values():
+      _maybe_add(sv.get("name"))
+
+  return list(strings)
```
Comment (pradh) [inline server/lib/translator.py:166]: I wonder if the logic would be simpler (without the need for `TranslationStringTokens`), if we just stripped the `${date}`-like stuff in the returned strings (in _maybe_add), and again when checking the map (in _maybe_update)?

File: server/lib/nl/common/utterance.py [hunk 4]
```diff
@@ -174,4 +174,8 @@ class Utterance:
   # Past complete context for insight flow.
   insight_ctx: Dict = field(default_factory=dict)
   # Test param.
-  test: str = ''
\ No newline at end of file
+  test: str = ''
+  # i18n param.
```
Comment (pradh) [inline server/lib/nl/common/utterance.py:178]: Wonder if we should have just 1 field `i18n_lang`?

File: server/routes/explore/api.py [hunk 5]
```diff
@@ -160,6 +160,7 @@ def _fulfill_with_insight_ctx(request: Dict, debug_logs: Dict,
                               counters: ctr.Counters) -> Dict:
   insight_ctx = request.get_json()
   test = request.args.get(Params.TEST.value, '')
+  i18n = request.args.get(Params.I18N.value, '')
```
Comment (pradh) [inline server/routes/explore/api.py:163]: This flow is for `/fulfill`, and it never sets the lang.  So I think if we want to support i18n for this flow, we should pass in `i18n_lang` as param.  In this PR, maybe drop changes to this file, and have a TODO to support i18n for `/fulfill` API?

File: server/routes/nl/helpers.py [hunk 6]
```diff
@@ -47,11 +47,15 @@
...
 from server.lib.translator import detect_lang_and_translate
+from server.lib.translator import translate_page_config
 from server.lib.util import get_nl_disaster_config
 from server.routes.nl import helpers
 import server.services.bigtable as bt
 import shared.lib.utils as shared_utils
 
+# English language code.
+_EN_LANG = "en"
+
```
Comment (pradh) [inline server/routes/nl/helpers.py:58]: move it to constants.py?

**Review by pradh — COMMENTED** [review]
Nicely done!

File: server/lib/translator.py (hunk 1 above)
Comment (keyurva) [reply server/lib/translator.py:68]: Done.

File: server/lib/translator.py (hunk 2 above)
Comment (keyurva) [reply server/lib/translator.py:90]: Done.

File: server/lib/nl/common/utterance.py (hunk 4 above)
Comment (keyurva) [reply server/lib/nl/common/utterance.py]: Done.

File: server/routes/explore/api.py (hunk 5 above)
Comment (keyurva) [reply server/routes/explore/api.py]: Done.

File: server/routes/nl/helpers.py (hunk 6 above)
Comment (keyurva) [reply server/routes/nl/helpers.py]: Done.

**Review by pradh — APPROVED** [review]
Thanks for the tests!

**Pull request merged**